Implement patch method

aljazkosir opened the pull request:
Implemented patch method that updates resource by merging request body and existing body.

miha-plesko commented:
I agree with @tadeboro, except regarding statuses where standard says [200 should be returned if at least one item was updatable else 405](http://redfish.dmtf.org/schemas/DSP0266_1.6.1.html#update-patch-a-id-update-patch-a-). Wdyt @tadeboro should we try to keep the mock 100% aligned with standard or should we adjust it a little to go with the actual state? I kind of think it'd be best to go with 100% standard-compliant, but I'm not sure if that's usable then.

tadeboro commented:
> Wdyt @tadeboro should we try to keep the mock 100% aligned with standard or should we adjust it a little to go with the actual state? I kind of think it'd be best to go with 100% standard-compliant, but I'm not sure if that's usable then.

It would be perfect if we could make the server fully Redfish standard compliant. But I am not sure how high on our list of priorities this compliance should be. If the Redfish server is only used for testing the clients, then full compliance is not that critical, since those same clients will need to work with the Redfish implementations that can be found in the wild.

aljazkosir commented:
@miha-plesko, @tadeboro I updated PR with the mentioned requirements. In cases when user wants to modify a structure of a resource (as in example above with `{"Bios": "Value"}`), 405 status code is returned. Similarly when user tries to create a new key 405 status code is returned. I also added some error messages to help user understand this status code better.

aljazkosir commented on lib/redfish_tools/exceptions.rb:
```diff
@@ -0,0 +1,5 @@
+require 'redfish_tools/exceptions'
```
I think I was trying to copy the `require` statement to `utils.rb`, but it somehow ended here :P

aljazkosir commented on lib/redfish_tools/servlet.rb:
```diff
@@ -78,8 +80,16 @@ def do_PUT(_request, response)
+      system = datastore.get(request.path).body
+      return response.status = 404 unless system
+
+      datastore.set(request.path.chomp("/"), combine_hashes(system, JSON.parse(request.body)))
```
I modified the `combine_hashes` function to copy the original hash and modifies the copy. Because currently if you do PATCH with payload
```
{
   {"SKU": "1234567"},
   {"Bios": "Value"}
}
```
405 status will be thrown (because if the `Bios` key) and `SKU` value will be updated (because it's processed before `Bios`). We probably don't want it to work like this.

tadeboro commented on lib/redfish_tools/utils.rb:
```diff
@@ -0,0 +1,24 @@
+
+module RedfishTools
+  module Utils
+    def Utils::combine_hashes(original, b, path: nil)
```
It would be more common to write this as `def self.combine_hashes(...)`. Then we can call this method like `Utils.combine_hashes(...)` from other modules/classes in Redfish tools code.

tadeboro commented on lib/redfish_tools/utils.rb:
```diff
@@ -0,0 +1,24 @@
+      b.each {|key, value|
+        if a.include?(key)
+          if (a[key].is_a? Hash) && (b[key].is_a? Hash)
+            a[key] = Utils::combine_hashes(a[key], b[key], path: path + [key.to_s])
```
We can drop the `Utils::` part when we redefine the `combine_hashes`.

aljazkosir commented:
@xlab-si/steampunk hello
